File: Java/src/com/ouroboros/include/Node.java
package com.ouroboros.include;

import java.util.Iterator;

public class Node implements Iterable<Node> {
    public int x, y;
    public Node next;

    public Node(int x, int y, Node next) {
        this.x = x;
        this.y = y;
        this.next = next;
    }

    @Override
    public Iterator<Node> iterator() {
        return new Iterator<Node>() {
            private Node curr = Node.this;

            @Override
            public boolean hasNext() {
                return curr != null;
            }

            @Override
            public Node next() {
                if (!hasNext())
                    throw new java.util.NoSuchElementException();
                Node tmp = curr;
                curr = curr.next;
                return tmp;
            }
        };
    }
}

File: Kotlin/app/src/main/kotlin/org/example/node.kt
package org.example

import java.util.NoSuchElementException

data class Node(
    var x: Int,
    var y: Int, 
    var next: Node? = null
) : Iterable<Node> {
    override fun iterator(): Iterator<Node> = object : Iterator<Node> {
        var curr: Node? = this@Node

        override fun hasNext(): Boolean = curr != null

        override fun next(): Node {
            val res = curr ?: throw NoSuchElementException()
            curr = res.next
            return res
        }
    }
}